lib/core/utils/get-selector.js

```diff
@@ -1,38 +1,179 @@
+
+  // Initialize the return structure with the three maps
+  let data = {
+    classes: {},
+    tags: {},
+    attributes: {}
+  };
+  
+  domTree = Array.isArray(domTree) ? domTree : [domTree];
+  let currentLevel = domTree.slice();
+  let stack = [];
+  while (currentLevel.length) {
+    let current = currentLevel.pop();
+    let node = current.actualNode;
+
+    if (!!node.querySelectorAll) { // ignore #text nodes
+
+      // count the tag
+      let tag = node.nodeName;
+      if (data.tags[tag]) {
+        data.tags[tag]++;
+      } else {
+        data.tags[tag] = 1;
+      }
+
+      // count all the classes
+      if (node.classList) {
+        Array.from(node.classList).forEach((cl) => {
+          const ind = escapeSelector(cl);
+          if (data.classes[ind]) {
+            data.classes[ind]++;
+          } else {
+            data.classes[ind] = 1;
+          }
+        });
+      }
+
+      // count all the filtered attributes
+      if (node.attributes) {
+        Array.from(node.attributes).filter(filterAttributes).forEach((at) => {
+          let atnv = getAttributeNameValue(node, at);
+          if (atnv) {
+            if (data.attributes[atnv]) {
+              data.attributes[atnv]++;
+            } else {
+              data.attributes[atnv] = 1;
+            }
+          }
+        });
+      }
+    }
+    if (current.children.length) {
+      // "recurse"
+      stack.push(currentLevel);
+      currentLevel = current.children.slice();
     }
-  }, Array.from(elm.classList).filter(isUncommonClassName));
+    while (!currentLevel.length && stack.length) {
+      currentLevel = stack.pop();
+    }
+  }
+  return data;
+};
+
+/**
+ * Given a node and the statistics on class frequency on the page,
+ * return all its uncommon class data sorted in order of decreasing uniqueness
+ * @param {Element} node      The node
+ * @param {Object} classData  The map of classes to counts
+ * @return {Array}            The sorted array of uncommon class data
+ */
+function uncommonClasses(node, selectorData) {
+  // jshint loopfunc:true
+  let retVal = [];
+  let classData = selectorData.classes;
+  let tagData = selectorData.tags;
+
+  if (node.classList) {
+    Array.from(node.classList).forEach((cl) => {
+      let ind = escapeSelector(cl);
+      if (classData[ind] < tagData[node.nodeName]) {
+        retVal.push({
+          name: ind,
+          count: classData[ind],
+          species: 'class'
+        });
+      }
+    });
+  }
+  return retVal.sort(countSort);
 }
 
-const commonNodes = [
-  'div', 'span', 'p',
-  'b', 'i', 'u', 'strong', 'em',
-  'h2', 'h3'
-];
+/**
+ * Given an element, return up to its three least common class names sorted in order of decreasing
+ * uniqueness
+ * @param {Element} elm           The Element
+ * @param {Object} selectorData   The statstics of the page
+ * @return {Array}                The sorted array of uncommon class names
+ */
+function getDistinctClassList (elm, selectorData) {
```

Comment:
This function is entirely redundant. `uncommonClasses` always returns an array, and `getThreeLeastCommonFeatures` already does the `.slice(0, 3)`. The same is true for `getDistinctAttributes`